// @corvu/utils/dist/chunk/7TEIQTJZ.js
import { tagName_default } from './A7SFFOAG.js';
import { isButton } from './2WOKOHZS.js';
import { mergeRefs } from './OZCI4NDN.js';
import { createComponent, Dynamic as Dynamic$1, mergeProps } from 'solid-js/web';
import { splitProps, createMemo, untrack, createSignal } from 'solid-js';

var DEFAULT_DYNAMIC_ELEMENT = "div";
var Dynamic = (props) => {
  const [localProps, otherProps] = splitProps(props, ["as"]);
  const cached = createMemo(() => localProps.as ?? DEFAULT_DYNAMIC_ELEMENT);
  const memoizedDynamic = createMemo(() => {
    const component = cached();
    switch (typeof component) {
      case "function":
        return untrack(() => component(otherProps));
      case "string":
        return createComponent(Dynamic$1, mergeProps({
          component
        }, otherProps));
    }
  });
  return memoizedDynamic;
};
var Dynamic_default = Dynamic;

// src/dynamic/DynamicButton.tsx
var DEFAULT_DYNAMIC_BUTTON_ELEMENT = "button";
var DynamicButton = (props) => {
  const [ref, setRef] = createSignal(null);
  const [localProps, otherProps] = splitProps(props, ["ref", "type"]);
  const tagName = tagName_default({
    element: ref,
    fallback: DEFAULT_DYNAMIC_BUTTON_ELEMENT
  });
  const memoizedIsButton = createMemo(() => {
    return isButton(tagName(), localProps.type);
  });
  return createComponent(Dynamic_default, mergeProps({
    ref(r$) {
      var _ref$ = mergeRefs(setRef, localProps.ref);
      typeof _ref$ === "function" && _ref$(r$);
    },
    get type() {
      return memoizedIsButton() ? "button" : void 0;
    },
    get role() {
      return !memoizedIsButton() ? "button" : void 0;
    }
  }, otherProps));
};
var DynamicButton_default = DynamicButton;

export { DynamicButton_default, Dynamic_default };
