// @corvu/utils/dist/chunk/FPO5WMIW.jsx
import {
  noPointerEvents_default
} from "./A4QYJDWU.jsx";
import {
  outsidePointer_default
} from "./34F543WF.jsx";
import {
  escapeKeyDown_default
} from "./GVREBUB5.jsx";

// src/create/dismissible.ts
import { mergeProps } from "solid-js";
var createDismissible = (props) => {
  const defaultedProps = mergeProps(
    {
      dismissOnEscapeKeyDown: true,
      dismissOnOutsidePointer: true,
      closeOnOutsidePointerStrategy: "pointerup",
      noOutsidePointerEvents: true
    },
    props
  );
  escapeKeyDown_default({
    enabled: defaultedProps.dismissOnEscapeKeyDown,
    onEscapeKeyDown: (event) => {
      defaultedProps.onEscapeKeyDown?.(event);
      if (!event.defaultPrevented) {
        defaultedProps.onDismiss("escapeKey");
      }
    }
  });
  outsidePointer_default({
    enabled: defaultedProps.dismissOnOutsidePointer,
    strategy: defaultedProps.closeOnOutsidePointerStrategy,
    ignore: defaultedProps.dismissOnOutsidePointerIgnore,
    onPointer: (event) => {
      defaultedProps.onOutsidePointer?.(event);
      if (!event.defaultPrevented) {
        const ctrlLeftClick = event.button === 0 && event.ctrlKey === true;
        const isRightClick = event.button === 2 || ctrlLeftClick;
        if (isRightClick)
          return;
        defaultedProps.onDismiss("pointerOutside");
      }
    },
    element: defaultedProps.element
  });
  noPointerEvents_default({
    enabled: defaultedProps.noOutsidePointerEvents
  });
};
var dismissible_default = createDismissible;

export {
  dismissible_default
};
