// @kobalte/core/dist/chunk/YIR5Y3GP.js
import { Polymorphic } from './SNFWSEOJ.js';
import { __export } from './5ZKAE4VZ.js';
import { createComponent, mergeProps } from 'solid-js/web';
import { mergeDefaultProps } from '@kobalte/utils';
import { createUniqueId, splitProps } from 'solid-js';

// src/skeleton/index.tsx
var skeleton_exports = {};
__export(skeleton_exports, {
  Root: () => Skeleton,
  Skeleton: () => Skeleton2
});
function Skeleton(props) {
  const defaultId = `skeleton-${createUniqueId()}`;
  const mergedProps = mergeDefaultProps({
    visible: true,
    animate: true,
    id: defaultId
  }, props);
  const [local, others] = splitProps(mergedProps, ["style", "radius", "animate", "height", "width", "visible", "circle"]);
  return createComponent(Polymorphic, mergeProps({
    as: "div",
    role: "group",
    get ["data-animate"]() {
      return local.animate;
    },
    get ["data-visible"]() {
      return local.visible;
    },
    get style() {
      return {
        "border-radius": local.circle ? "9999px" : local.radius ? `${local.radius}px` : void 0,
        width: local.circle ? `${local.height}px` : local.width ? `${local.width}px` : "100%",
        height: local.height ? `${local.height}px` : "auto",
        ...local.style
      };
    }
  }, others));
}

// src/skeleton/index.tsx
var Skeleton2 = Skeleton;

export { Skeleton, Skeleton2, skeleton_exports };
